engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/UpdateByWindow.java

```diff
@@ -0,0 +1,404 @@
+            context.inputSourceChunkPopulated[srcIdx] = true;
+        }
+    }
+
+    /**
+     * Perform the computations and store the results in the operator output sources
+     *
+     * @param context the window context that will manage the results.
+     * @param initialStep whether this is the creation step of the table.
+     */
+    public abstract void processRows(final UpdateByWindowContext context, final boolean initialStep);
+
+    /**
+     * Returns `true` if the window for this bucket needs to be processed this cycle.
+     *
+     * @param context the window context that will manage the results.
+     */
+    public boolean isWindowDirty(final UpdateByWindowContext context) {
+        return context.isDirty;
+    }
+
+    /**
+     * Returns `true` if the window for this bucket needs to be processed this cycle.
+     *
+     * @param context the window context that will manage the results.
+     */
+    public int[] getDirtyOperators(final UpdateByWindowContext context) {
+        return context.dirtyOperatorIndices;
+    }
+
+    /**
+     * Returns the list of input sources that will be needed to process the `dirty` operators for this bucket
+     *
+     * @param context the window context that will manage the results.
+     */
+    public int[] getDirtySources(final UpdateByWindowContext context) {
+        return context.dirtySourceIndices;
+    }
+
+    /**
+     * Returns the rows that will be recomputed for this bucket this cycle
+     *
+     * @param context the window context that will manage the results.
+     */
+    public RowSet getAffectedRows(final UpdateByWindowContext context) {
+        return context.affectedRows;
+    }
+
+    /**
+     * Returns the rows that are needed to recompute the `affected` rows (that `influence` the results)
+     *
+     * @param context the window context that will manage the results.
+     */
+    public RowSet getInfluencerRows(final UpdateByWindowContext context) {
+        return context.influencerRows;
+    }
+
+    // endregion
+
+    /**
+     * Returns a hash code to help distinguish between windows on the same UpdateBy call
+     */
+    protected static int hashCode(boolean windowed, @NotNull String[] inputColumnNames,
+            @Nullable String timestampColumnName, long prevUnits,
+            long fwdUnits) {
+
+        // hash the input column names
+        int hash = 0;
+        for (String s : inputColumnNames) {
+            hash = 31 * hash + s.hashCode();
+        }
+
+        // treat all cumulative ops with the same input columns as identical, even if they rely on timestamps
+        if (!windowed) {
+            return 31 * hash + Boolean.hashCode(false);
+        }
+
+        // windowed ops are unique per type (ticks/time-based) and window dimensions
+        hash = 31 * hash + Boolean.hashCode(true);
+        hash = 31 * hash + Boolean.hashCode(timestampColumnName != null);
```

Comment:
Future proof this with:
```suggestion
        hash = 31 * hash + Objects.hashCode(timestampColumnName);
```